text/000-rustdoc-cfgs-handling.md

```diff
@@ -0,0 +1,283 @@
+
+```rust
+#[doc(cfg(any(windows, unix)))]
+pub mod desktop {
+    #[doc(cfg(not(unix)))]
+    pub mod non_unix {
+        //
+    }
+}
+```
+
+> ![Available on (Windows or Unix) and non-Unix only.](https://hackmd.io/_uploads/SJrmwYeF2.png)
+
+[Future versions of rustdoc][boolean simplification] may simplify this display down to "available on **Windows** only."
+
+### Re-exports and inlining
+
+`cfg` attributes of a re-export are never merged the re-exported item(s). If `#[doc(inline)]` attribute is used on a re-export, the `cfg` of the re-exported item will be merged with the re-export's.
+
+```rust
+#[doc(cfg(any(windows, unix)))]
+pub mod desktop {
+    #[doc(cfg(not(unix)))]
+    pub mod non_unix {
+        //
+    }
+}
+
+#[doc(cfg(target_os = "freebsd"))]
+pub use desktop::non_unix as non_unix_desktop;
+#[doc(cfg(target_os = "macos"))]
+#[doc(inline)]
+pub use desktop::non_unix as inlined_non_unix_desktop;
+```
+
+In this example, `non_unix_desktop` will only display `cfg(target_os = "freeebsd")` and not display any `cfg` from `desktop::non_unix`.
+
+On the contrary, `inlined_non_unix_desktop` will have cfgs from both the re-export and the re-exported item.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+A potential drawback is that it adds more attributes, making documentation more complex.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Why not merging cfg and doc(cfg) attributes by default?
+
+It was debated and implemented in [rust-lang/rust#113091](https://github.com/rust-lang/rust/pull/113091).
+
+When re-exporting items with different cfgs there are two things that can happen:
+
+ 1. The re-export uses a subset of cfgs, this subset is sufficient so that the item will appear exactly with the subset
+ 2. The re-export uses a non-subset of cfgs like in this code:
+   ```rust
+    #![feature(doc_auto_cfg)]
+
+    #[cfg(target_os = "linux")]
+    mod impl_ {
+        pub fn foo() { /* impl for linux */ }
+    }
+
+    #[cfg(target_os = "macos")]
+    mod impl_ {
+        pub fn foo() { /* impl for darwin */ }
+    }
+
+    pub use impl_::foo;
+   ```
+   If the non-subset cfgs are active (e.g. compiling this example on windows), then this will be a compile error as the item doesn't exist to re-export. If the subset cfgs are active it behaves like described in 1.
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+
+# Future possibilities
+[future possibilities]: #future-possibilities
+
+## Boolean simplification
+[boolean simplification]: #boolean-simplification
+
+> ![Available on (Windows or Unix) and non-Unix only.](https://hackmd.io/_uploads/SJrmwYeF2.png)
+
+Of course, the above example is equivalent to "available on **Windows** only."
+
+Making this actually work all the time is equivalent to a [boolean satisfiability] check, coliquially called a "SAT problem," and can take exponential time.
```

Comment:
This is… actually not true. The format you'd expect is probably akin to [dysjunctive normal form](https://en.wikipedia.org/wiki/Disjunctive_normal_form) or [conjunctive normal form](https://en.wikipedia.org/wiki/Conjunctive_normal_form) and converting to either of these is a polynomial-time algorithm.

I wouldn't expect this to be a requirement to stabilise, but, just wanted to correct the statement that this would take exponential time. Simplifying expressions actually is polynomial time; the exponential-time algorithm is figuring out the cases where the `cfg` could actually be applied, which is beyond the scope of rustdoc.

Reply:
It's a very easy mistake to make! Unfortunately it's not a common part of maths or even computer science education, but it is very useful for simplifying boolean expressions.

Reply:
Thanks for the explanations, it's very welcome!

Reply:
I removed the technical part as I think it's out of context here. The linked issue remains so the discussion can continue there. :)